src/sync/PushManager/index.js

```diff
@@ -0,0 +1,201 @@
+      reauthBackoff.scheduleCall() : // reauthenticate in case of token expired (when somehow refresh token was not properly executed) or invalid
+      sseReconnectBackoff.scheduleCall(); // reconnect SSE for any other SSE error
+
+    const errorMessage = error.parsedData && error.parsedData.message;
+    log.error(`Fail to connect to streaming${errorMessage ? `, with error message: "${errorMessage}"` : ''}. Attempting to reconnect in ${delayInMillis / 1000} seconds.`);
+
+    pushEmitter.emit(PUSH_DISCONNECT); // no harm if polling already
+  });
+
+  /** Functions related to synchronization (Queues and Workers in the spec) */
+
+  const producer = context.get(context.constants.PRODUCER);
+  const splitUpdateWorker = new SplitUpdateWorker(storage.splits, producer);
+
+  pushEmitter.on(SPLIT_KILL, splitUpdateWorker.killSplit.bind(splitUpdateWorker));
+  pushEmitter.on(SPLIT_UPDATE, splitUpdateWorker.put.bind(splitUpdateWorker));
+
+  if (clientContexts) { // browser
+    pushEmitter.on(MY_SEGMENTS_UPDATE, function handleMySegmentsUpdate(parsedData, channel) {
+      const userKeyHash = channel.split('_')[2];
+      const userKey = userKeyHashes[userKeyHash];
+      if (userKey && clientContexts[userKey]) { // check context since it can be undefined if client has been destroyed
+        const mySegmentSync = clientContexts[userKey].get(context.constants.MY_SEGMENTS_CHANGE_WORKER, true);
+        mySegmentSync && mySegmentSync.put(
+          parsedData.changeNumber,
+          parsedData.includesPayload ? parsedData.segmentList ? parsedData.segmentList : [] : undefined);
+      }
+    });
+  } else { // node
+    const segmentUpdateWorker = new SegmentUpdateWorker(storage.segments, producer);
+    pushEmitter.on(SEGMENT_UPDATE, segmentUpdateWorker.put.bind(segmentUpdateWorker));
+  }
+
+  return Object.assign(
+    // Expose Event Emitter functionality and Event constants
+    Object.create(pushEmitter),
+    {
+
+      // Expose functionality for starting and stoping push mode:
+
+      stop: disconnectPush,
+
+      // used in node
+      start: connectPush,
+
+      // used in browser
+      startNewClient(userKey, context) {
+        const hash = hashUserKey(userKey);
+        if (!userKeyHashes[hash]) {
+          userKeyHashes[hash] = userKey;
+          connectForNewClient = true; // we must reconnect on start, to listen the channel for the new user key
+        }
+        const storage = context.get(context.constants.STORAGE);
+        const producer = context.get(context.constants.PRODUCER);
+        context.put(context.constants.MY_SEGMENTS_CHANGE_WORKER, new SegmentUpdateWorker(storage.segments, producer));
+
+        // Reconnects in case of a new client.
+        // Run in next event-loop cycle to save authentication calls
+        // in case the user is creating several clients in the current cycle.
+        setTimeout(function start() {
```

Comment:
If you give it name for debugging (which is good, as it'll show up on stack traces) make it a bit clearer. maybe checkForReconnect ?